### DISCUSSION TITLE
Disable telemetry in next.config.json

### DISCUSSION THREAD

**Discussion by opiation** [vercel/next.js#31400] [Ideas]
### Describe the feature you'd like to request

I want to disable Nextjs telemetry in a version-controlled way.

### Describe the solution you'd like

Since most other Nextjs options are conventionally configured in `next.config.json`, this seems like the most appropriate place to support such a setting.

### Describe alternatives you've considered

Using the environment variable `NEXT_TELEMETRY_DISABLED=1` is not adequate since _conventionally_, environment variables or `.env` files are used to store sensitive project secrets so they are not under version control, thus requiring manual intervention from every project member.

The workaround I currently use is to add the following to `package.json`'s `scripts`:

```json
{ "postinstall": "yarn run next telemetry disable" }
```

This works but it's an unnecessary step after `yarn install` has been run once on a given machine.

> Note, I would prefer if telemetry was disabled by default but it doesn't seem like Vercel is willing to accept that request from the community given the numerous closed issues on the matter so I won't bother with a PR making such a change. Using a tool to help me develop applications on my local machine, I would never expect this tool to report telemetry to some remote server without asking me first, regardless of how _anonymous_ the telemetry is (yes, even if the git origin remote URL is [one-way hashed][0], even if you can't uniquely identify me from [my machine specs and env][1]). One can geolocate the telemetry requests and get pretty close in my experience anyways. Ultimately, external network communication from my machine just isn't your call. :shrug:

[0]: https://github.com/vercel/next.js/blob/d9fc3eeb7e7fc4e9e35feaabacb33b790a2705da/packages/next/telemetry/storage.ts#L162
[1]: https://github.com/vercel/next.js/blob/canary/packages/next/telemetry/anonymous-meta.ts#L30-L49

**Comment by will-stone** [comment]
Hi @opiation Next [recommend](https://nextjs.org/docs/basic-features/environment-variables#default-environment-variables) committing `.env` to version control. It's only `.env.*.local` that should be gitignored.

I've put `NEXT_TELEMETRY_DISABLED=1` in `.env`, do you know if there's a way to check this has worked?

**Reply by will-stone** [reply]
I checked using Little Snitch and can confirm `NEXT_TELEMETRY_DISABLED=1` inside `.env` works just fine 🙂

**Reply by opiation** [reply]
Hey @will-stone, thanks for chiming in. I'm aware that the `NEXT_TELEMETRY_DISABLED` env var works this way. I realize that Next recommends committing the `.env` to the repository but that's not a decision for Next to make for us. I understand that environment variables used by my Next application will end up on the end user's machine by virtue of these being Javascript browser applications. That is not true of all environment variables however, only of those that are actually used by the Javascript application. If I choose for my project repository to include both web client and related API server with settings in a single `.env` (for instance), that `.env` should remain private as it will contain sensitive keys only the server should know about. I'd rather not check all my application's sensitive API keys under version control, especially not justified as a convenient way to disable Next's telemetry that really should have been disabled by default in the first place.

In addition, Next's config JSON file is a rather obvious place to supply such non-sensitive configurations. The responsibility of this file is **only** to Next while `.env` may be used by many other applications, libraries and tools holding both sensitive and non-sensitive information. I hope that clarifies my reasons for now wanting to use the environment variable to disable telemetry.

**Reply by will-stone** [reply]
I agree that this should be in Next config.  In the meantime, note that only env vars prefixed with NEXT_PUBLIC_ can make it to the client-side bundle. Then for secrets you can use the .env.*.locals files. Also agree that the telemetry should be opt-in, but then there’s a lot we get for free that I see setting this env var, and committing it to git, a small price to pay.

**Reply by akomm** [reply]
@opiation

This is why next supports .env.loca .env.dev.local, etc.

You have .env with all the safe things and defaults and .local unversioned.

But I agree having the config option would be better. There is no easy documented way to figure out if it even works, besides maybe tracking network traffic. Using the NEXT_TELEMETRY_DEBUG=1 option will always show  the data, regardless of telemetry setting. This might not mean the data is sent, but it leaves it open to you to guess or analyze traffic.

Feels a bit like if its done that way so maybe someone in team forgets to disable it.

We all know the 100% anonymous claim is always false. Netflix has claimed that once and made anonymous data public, which turned out to be not so anonymous once you have other datasets to heuristically analyze and connect them by patterns.

And on top of this while I can see in debug what is sent, I can't ever know all the data sent in which case as its undocumented and subject to change any moment (https://nextjs.org/telemetry)

Actually one of the reasons for me to disable it is this whole constructed intransparency of whether it actually runs or not when using env and what data in which case is sent as well as the lack of notification when it changes (nobody is going to check the site every time they update next)

**Comment by MariuzM** [comment]
I'm testing this and does this mean telemetriy is enabled? 

```
 [telemetry] {
   "eventName": "NEXT_CLI_SESSION_STARTED",
   "payload": {
     "nextVersion": "13.4.19",
     "nodeVersion": "v18.17.0",
     "cliCommand": "dev",
     "isSrcDir": true,
     "hasNowJson": false,
     "isCustomServer": false,
     "hasNextConfig": true,
     "buildTarget": "default",
     "hasWebpackConfig": false,
     "hasBabelConfig": false,
     "imageEnabled": true,
     "imageFutureEnabled": true,
     "basePathEnabled": false,
     "i18nEnabled": false,
     "locales": null,
     "localeDomainsCount": null,
     "localeDetectionEnabled": null,
     "imageDomainsCount": 0,
     "imageRemotePatternsCount": 1,
     "imageSizes": "16,32,48,64,96,128,256,384",
     "imageLoader": "default",
     "imageFormats": "image/webp",
     "nextConfigOutput": null,
     "trailingSlashEnabled": false,
     "reactStrictMode": false,
     "webpackVersion": 5,
     "turboFlag": false,
     "appDir": true,
     "pagesDir": false
   }
 }
```

**Comment by tofran** [comment]
I want to add one unbelievable thing that Vercel is doing is cross-project tracking.

Meaning that the so called "anonymous" tracking sets a global  `anonymousId` that is persisted forever and used to track all actions on Vercel projects across all repositories.

Mine is stored in `/Users/USER/Library/Preferences/nextjs-nodejs/config.json`. From what I understand it uses some pre-build version of [sindresorhus/conf](https://github.com/sindresorhus/conf).

If you are distributing an open source project that uses Next, you should make sure you have the postinstall script, otherwise you may be violating EU law (I don't care about going against Next/Vercel as they are an American giant, but I do care that what I distribute is ethical and compliant - you should too).

Note: For the `postinstall` script, I recommend calling `next` directly as you are already in the node context (works regardless of the package manager):

```
{ "postinstall": "next telemetry disable" }
```
